```
./loader/cache/MemoryCache.cpp:177:40: error: assigning to 'WebCore::CachedResource *' from incompatible type 'WTF::KeyValuePair<std::pair<WTF::URL, WTF::String>, WTF::WeakPtr<WebCore::CachedResource>>::ValueType' (aka 'WTF::WeakPtr<WebCore::CachedResource>')
```

--- a/LayoutTests/http/tests/workers/memory-cache-crash-expected.txt
+++ b/LayoutTests/http/tests/workers/memory-cache-crash-expected.txt
@@ -1,0 +1,1 @@
+This test passes if it doesn't crash.

--- a/LayoutTests/http/tests/workers/memory-cache-crash.html
+++ b/LayoutTests/http/tests/workers/memory-cache-crash.html
@@ -1,0 +1,19 @@
+<p>This test passes if it doesn't crash.</p>
+<script>
+  if (window.testRunner) {
+    testRunner.dumpAsText();
+    testRunner.waitUntilDone();
+  }
+
+  onload = async () => {
+    document.createElement('audio').src = 'data:';
+    for (let i = 0; i < 1000; i++) {
+      new Worker('');
+      await navigator.storage.persist();
+    }
+    setTimeout(() => {
+      if (window.testRunner)
+        testRunner.notifyDone();
+    }, 50);
+  };
+</script>

--- a/Source/WebCore/loader/cache/CachedResource.cpp
+++ b/Source/WebCore/loader/cache/CachedResource.cpp
@@ -755,6 +755,17 @@ void CachedResource::setResourceToRevalidate(CachedResource* resource)
     m_resourceToRevalidate = resource;
 }
 
+void CachedResource::replaceResourceToRevalidate(CachedResource& resource)
+{
+    ASSERT(m_resourceToRevalidate);
+
+    m_resourceToRevalidate->m_proxyResource = nullptr;
+    m_resourceToRevalidate->deleteIfPossible();
+    m_resourceToRevalidate = &resource;
+    ASSERT(!resource.m_proxyResource);
+    resource.m_proxyResource = this;
+}
+
 void CachedResource::clearResourceToRevalidate() 
 {
     ASSERT(m_resourceToRevalidate);

--- a/Source/WebCore/loader/cache/CachedResource.h
+++ b/Source/WebCore/loader/cache/CachedResource.h
@@ -282,6 +282,7 @@ public:
 
     // HTTP revalidation support methods for CachedResourceLoader.
     void setResourceToRevalidate(CachedResource*);
+    void replaceResourceToRevalidate(CachedResource&);
     virtual void switchClientsToRevalidatedResource();
     void clearResourceToRevalidate();
     void updateResponseAfterRevalidation(const ResourceResponse& validatingResponse);

--- a/Source/WebCore/loader/cache/MemoryCache.cpp
+++ b/Source/WebCore/loader/cache/MemoryCache.cpp
@@ -146,9 +146,9 @@ void MemoryCache::revalidationSucceeded(CachedResource& revalidatingResource, co
     ASSERT(response.source() == ResourceResponse::Source::MemoryCacheAfterValidation);
     ASSERT(revalidatingResource.resourceToRevalidate());
     CachedResourceHandle protectedRevalidatingResource { revalidatingResource };
-    CachedResource& resource = *revalidatingResource.resourceToRevalidate();
-    ASSERT(!resource.inCache());
-    ASSERT(resource.isLoaded());
+    auto* resource = revalidatingResource.resourceToRevalidate();
+    ASSERT(!resource->inCache());
+    ASSERT(resource->isLoaded());
 
     // Calling remove() can potentially delete revalidatingResource, which we use
     // below. This mustn't be the case since revalidation means it is loaded
@@ -157,23 +157,31 @@ void MemoryCache::revalidationSucceeded(CachedResource& revalidatingResource, co
 
     remove(revalidatingResource);
 
-    auto& resources = ensureSessionResourceMap(resource.sessionID());
-    auto key = std::make_pair(resource.url(), resource.cachePartition());
+    auto& resources = ensureSessionResourceMap(resource->sessionID());
+    std::pair key { resource->url(), resource->cachePartition() };
 
-    RELEASE_ASSERT(!resources.get(key));
-    resources.set(key, &resource);
-    resource.setInCache(true);
-    resource.updateResponseAfterRevalidation(response);
-    insertInLRUList(resource);
-    long long delta = resource.size();
-    if (resource.decodedSize() && resource.hasClients())
-        insertInLiveDecodedResourcesList(resource);
-    if (delta)
-        adjustSize(resource.hasClients(), delta);
+    auto addResult = resources.add(key, resource);
+    if (addResult.isNewEntry) {
+        resource->setInCache(true);
+        resource->updateResponseAfterRevalidation(response);
+        insertInLRUList(*resource);
+        long long delta = resource->size();
+        if (resource->decodedSize() && resource->hasClients())
+            insertInLiveDecodedResourcesList(*resource);
+        if (delta)
+            adjustSize(resource->hasClients(), delta);
+    } else {
+        // The resource was re-inserted in the cache during its revalidation.
+        // Ignore the revalidated resource and switch clients to the one that
+        // is already in the cache.
+        resource = addResult.iterator->value.get();
+        ASSERT(resource);
+        revalidatingResource.replaceResourceToRevalidate(*resource);
+    }
 
     revalidatingResource.switchClientsToRevalidatedResource();
     ASSERT(!revalidatingResource.m_deleted);
-    // this deletes the revalidating resource
+    // This deletes the revalidating resource.
     revalidatingResource.clearResourceToRevalidate();
 }
 

